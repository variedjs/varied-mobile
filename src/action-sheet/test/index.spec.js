import { mount, shallowMount } from "@vue/test-utils";
import ActionSheet from "..";
import { later } from "../../../test";

test("should emit select event after clicking option", () => {
  const wrapper = mount(ActionSheet, {
    propsData: {
      value: true,
      actions: [{ name: "Option" }]
    }
  });

  wrapper.find(".vm-action-sheet-item").trigger("click");
  expect(wrapper.emitted("select").length).toEqual(1);
  expect(wrapper.emitted("select")[0][0]).toEqual({ name: "Option" });
});

test("should call callback function after clicking option", () => {
  const callback = jest.fn();
  const wrapper = mount(ActionSheet, {
    propsData: {
      value: true,
      actions: [{ name: "Option", callback }]
    }
  });

  wrapper.find(".vm-action-sheet-item").trigger("click");
  expect(callback).toHaveBeenCalledTimes(1);
});

test("should not emit select event after clicking loading option", () => {
  const wrapper = mount(ActionSheet, {
    propsData: {
      value: true,
      actions: [{ name: "Option", loading: true }]
    }
  });

  wrapper.find(".vm-action-sheet-item").trigger("click");
  expect(wrapper.emitted("select")).toBeFalsy();
});

test("should not emit select event after clicking disabled option", () => {
  const wrapper = mount(ActionSheet, {
    propsData: {
      value: true,
      actions: [{ name: "Option", disabled: true }]
    }
  });

  wrapper.find(".vm-action-sheet-item").trigger("click");
  expect(wrapper.emitted("select")).toBeFalsy();
});

test("should emit cancel event after clicking cancel button", () => {
  const wrapper = mount(ActionSheet, {
    propsData: {
      value: true,
      actions: [{ name: "Option" }],
      cancelText: "Cancel"
    }
  });

  wrapper.find(".vm-action-sheet-cancel").trigger("click");
  expect(wrapper.emitted("cancel").length).toEqual(1);
});

test("should render subname correctly", () => {
  const wrapper = mount(ActionSheet, {
    propsData: {
      value: true,
      actions: [{ name: "Option", subname: "Subname" }],
      cancelText: "Cancel"
    }
  });

  expect(wrapper.find(".vm-action-sheet-item").html()).toMatchSnapshot();
});

test("should render content after disabling the lazy-render prop", async () => {
  const wrapper = mount(ActionSheet);
  expect(wrapper.find(".vm-action-sheet-content").exists()).toBeFalsy();

  await wrapper.setProps({ lazyRender: false });
  expect(wrapper.find(".vm-action-sheet-content").exists()).toBeTruthy();
});

test("should render default slot correctly", () => {
  const wrapper = shallowMount(ActionSheet, {
    props: {
      value: true,
      title: "Title"
    },
    slots: {
      default: `<div>内容</div>`
    }
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('should have "vm-popup-round" class when setting the round prop', () => {
  const wrapper = mount(ActionSheet, {
    propsData: {
      value: true,
      round: true
    }
  });

  expect(wrapper.find(".vm-popup-round").exists()).toBeTruthy();
});

test("should change option color when using the color prop", () => {
  const wrapper = mount(ActionSheet, {
    propsData: {
      value: true,
      actions: [{ name: "Option", color: "red" }]
    }
  });

  expect(wrapper.find(".vm-action-sheet-item").element.style.color).toEqual(
    "red"
  );
});

test("should hide close icon when the closeable prop is false", async () => {
  const wrapper = mount(ActionSheet, {
    propsData: {
      value: true,
      title: "Title"
    }
  });

  expect(wrapper.find(".vm-action-sheet-close").exists()).toBeTruthy();

  await wrapper.setProps({ closeable: false });
  expect(wrapper.find(".vm-action-sheet-close").exists()).toBeFalsy();
});

test("should allow to custom close icon with closeIcon prop", () => {
  const wrapper = mount(ActionSheet, {
    propsData: {
      value: true,
      title: "Title",
      closeIcon: "cross"
    }
  });

  expect(wrapper.find(".vm-action-sheet-close").html()).toMatchSnapshot();
});

test("should render description correctly", () => {
  const wrapper = mount(ActionSheet, {
    propsData: {
      value: true,
      description: "This is a description"
    }
  });

  expect(wrapper.find(".vm-action-sheet-description").html()).toMatchSnapshot();
});

test("should render description slot when match snapshot", () => {
  const wrapper = shallowMount(ActionSheet, {
    props: {
      value: true
    },
    slots: {
      description: `<div>Custom Description</div>`
    }
  });

  expect(wrapper.find(".vm-action-sheet-description").html()).toMatchSnapshot();
});

test("should close after clicking option if close-on-click-action prop is true", () => {
  const wrapper = mount(ActionSheet, {
    propsData: {
      value: true,
      actions: [{ name: "Option" }],
      closeOnClickAction: true
    }
  });

  const option = wrapper.find(".vm-action-sheet-item");
  option.trigger("click");

  expect(wrapper.emitted("input").length).toEqual(1);
  expect(wrapper.emitted("input")[0][0]).toEqual(false);
});
