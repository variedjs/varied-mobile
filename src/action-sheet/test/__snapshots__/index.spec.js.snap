// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow to custom close icon with closeIcon prop 1`] = `
<i class="vm-icon vm-icon-cross vm-action-sheet-close">
  <!---->
</i>
`;

exports[`should render default slot correctly 1`] = `
<popup-stub overlay="true" closeonclickoverlay="true" lockscroll="true" lazyrender="true" round="true" closeicon="close" closeiconposition="top-right" position="bottom" safeareainsetbottom="true" class="vm-action-sheet">
  <!---->
  <!---->
  <div class="vm-action-sheet-content">
    <div>内容</div>
  </div>
  <!---->
</popup-stub>
`;

exports[`should render description correctly 1`] = `
<div class="vm-action-sheet-description">
  This is a description
</div>
`;

exports[`should render description slot when match snapshot 1`] = `
<div class="vm-action-sheet-description">
  <div>Custom Description</div>
</div>
`;

exports[`should render subname correctly 1`] = `
<button type="button" class="vm-action-sheet-item"><span class="vm-action-sheet-name">Option</span>
  <div class="vm-action-sheet-subname">
    Subname
  </div>
</button>
`;
